### PR TITLE
lib: Added nrf9160_timestamp module

### PR DESCRIPTION
**Pull request by simensrostad**
This module maintains a date time in UTC from multiple sources that can
be fetched in order to timestamp data in UNIX time ms.

Signed-off-by: Simen S. Røstad <simen.rostad@nordicsemi.no>

### CODE REVIEW

**Comment by rlubos** [comment]
Doesn't the functionality in here partially overlap with what we have in https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/posix/clock.c? IMO we should not re-invent the wheel.

**Comment by simensrostad** [comment]
@rlubos Yes, it looks like that particular library can be used to keep track of an internal system time, much like what is already implemented. I tried to add the suggested library but was unable to verify if it works or not because of build errors seemingly related to issue https://github.com/zephyrproject-rtos/zephyr/issues/12368. In my understanding the posix subsys has poor coherence with bsd lib at the moment.

File: lib/nrf9160_timestamp/nrf9160_timestamp.c [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+void date_time_set(struct tm *new_date_time)
+{
+	time_aux.last_date_time_update = k_uptime_get();
+
+	/* Needed changes so mktime gives UNIX time */
+	new_date_time->tm_year    -= 1900;
+	new_date_time->tm_mon     -= 1;
+
+	time_aux.date_time_utc =
+		(s64_t)mktime(new_date_time) * 1000 - k_uptime_get();
```
Comment (pabigot) [inline lib/nrf9160_timestamp/nrf9160_timestamp.c]: `mktime()` converts times taking into account the time zone; it does not guarantee to produce UTC.  Consider using the Zephyr function `timeutil_timegm64()` from the upstream `<sys/timeutil.h>` header.

File: lib/nrf9160_timestamp/nrf9160_timestamp.c [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	date_time.tm_hour = parse_time_entries(buf, 17, 18);
+	date_time.tm_min  = parse_time_entries(buf, 20, 21);
+	date_time.tm_sec  = parse_time_entries(buf, 23, 24);
+
+	if (date_time.tm_year == 115 || date_time.tm_year == 0) {
+		LOG_DBG("Modem giving old cellular network time");
+		return -ENOMSG;
+	}
+
+	time_aux.date_time_utc =
```
Comment (pabigot) [inline lib/nrf9160_timestamp/nrf9160_timestamp.c]: Use `timeutil_timegm64` (see other comment).

File: lib/nrf9160_timestamp/nrf9160_timestamp.c [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+	k_delayed_work_init(&time_work.work, nrf9160_time_handler);
+	k_delayed_work_submit(&time_work.work, K_NO_WAIT);
+}
+
+void date_time_set(struct tm *new_date_time)
+{
+	time_aux.last_date_time_update = k_uptime_get();
+
+	/* Needed changes so mktime gives UNIX time */
```
Comment (pabigot) [inline lib/nrf9160_timestamp/nrf9160_timestamp.c]: This makes no sense.  The API should assume the incoming `struct tm` follows the standard interpretation, so the year is already relative to 1900 CE and the month ranges from 0 to 11.

If this were going upstream I'd want to see a unit test that verifies that these functions work as intended, i.e. you get/set/get multiple times and time is properly adjusted.

Also this should be taking a `const struct tm *`: mutating the passed in timestamp is not a good idea.

File: lib/nrf9160_timestamp/nrf9160_timestamp.c [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+	if (age_last_time_update > current_update_time_interval_ms) {
+		LOG_DBG("Current date time too old");
+		return -ENODATA;
+	}
+
+	return 0;
+}
+
+void update_new_date_time(void)
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply lib/nrf9160_timestamp/nrf9160_timestamp.c]: Done

File: lib/nrf9160_timestamp/Kconfig [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	select AT_CMD
+	select SNTP
+
+if NRF9160_TIMESTAMP
+
+config NRF9160_TIMESTAMP_TIME_UPDATE_INTERVAL
+	int "Time update interval"
+	default 3600
+
+config NRF9160_TIMESTAMP_DEV_NAME
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply lib/nrf9160_timestamp/Kconfig]: No, it is not necessary. It was simply residue from an earlier implementation.

File: lib/nrf9160_timestamp/nrf9160_timestamp.c (hunk 3 above)
Comment (simensrostad) [reply lib/nrf9160_timestamp/nrf9160_timestamp.c]: Removed the convertion and changed the passing argument to const.

File: lib/nrf9160_timestamp/nrf9160_timestamp.c (hunk 1 above)
Comment (simensrostad) [reply lib/nrf9160_timestamp/nrf9160_timestamp.c]: Agreed. Changed to ``timeutil_timegm64``

File: include/nrf9160_timestamp.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ *         This function requires that k_uptime_get has been called
+ *         on the passing variable unix_timestamp_ms variable at a point
+ *         prior to calling get_date_time.
+ *
+ *  @param timestamp Pointer to the timestamp structure.
+ *
+ *  @return 0 If the operation was successful.
+ *            Otherwise, a (negative) error code is returned.
+ */
+int date_time_get(s64_t *unix_timestamp_ms);
```
Comment (jhn-nordic) [inline include/nrf9160_timestamp.h]: I suggest that you have two get functions in the API. On for getting the current time now that is called date_time_get()
and one that is called something that explains that it get date time at a point back in time. (like this funciton works now)

File: lib/nrf9160_timestamp/nrf9160_timestamp.c [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+{
+	k_delayed_work_init(&time_work, nrf9160_time_handler);
+	k_delayed_work_submit(&time_work, K_NO_WAIT);
+}
+
+void date_time_set(const struct tm *new_date_time)
+{
+	time_aux.last_date_time_update = k_uptime_get();
+	time_aux.date_time_utc =
+		(s64_t)timeutil_timegm64(new_date_time) * 1000 - k_uptime_get();
```
[the remark this answers is not in the document]
Comment (jhn-nordic) [reply lib/nrf9160_timestamp/nrf9160_timestamp.c]: Is this the function you can ie use to feed it with GPS time?

[the remark this answers is not in the document]
Comment (jhn-nordic) [reply lib/nrf9160_timestamp/nrf9160_timestamp.c]: Would it make sense to restart the time_work delayed work timer? No need to fetch time from other sources if you feed it with a reliable source.

**Comment by simensrostad** [comment]
This PR has been moved to another fork of NCS.  #1861

**Pull request closed without merging**